### PR TITLE
Implement native support StringViewArray for `regexp_is_match` and `regexp_is_match_scalar` function, deprecate `regexp_is_match_utf8` and `regexp_is_match_utf8_scalar`

### LINKED ISSUE #6370: implement `regexp_is_match_utf8` and `regexp_is_match_utf8_scalar` for `StringViewArray`
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
We have implemented support for  `StringViewArray` in arrow-rs (see  https://github.com/apache/arrow-rs/issues/6163 and https://www.influxdata.com/blog/faster-queries-with-stringview-part-one-influxdb/ ) and to take full advantage of it, kernels that operate on strings (`StringArray` and `LargeStringArray`) need to be updated with specialized implementations that can
1. Natively operate on `StringViewArray` without having to covert first to `StringArray`
2. (Potentially) take advantage of the new string view layout 

`regexp_is_match_utf8` is one of these functions that does not have native `StringViewArray` support yet
 
**Describe the solution you'd like**
1. Implement [`regexp_is_match_utf8` ](https://docs.rs/arrow/latest/arrow/compute/fn.regexp_is_match_utf8.html)for `StringViewArray`
2. Implement [`regexp_is_match_utf8_scalar`](https://docs.rs/arrow/latest/arrow/compute/fn.regexp_is_match_utf8_scalar.html) for `StringViewArray`
3. Tests for same

**Describe alternatives you've considered**
N/A

**Additional context**
@tlm365 implemented this downstream in datafusion --  https://github.com/apache/datafusion/pull/12168. We can probably take a friendly look / move the implementation up to arrow-rs

### PR DESCRIPTION
**Pull request by tlm365**
# Which issue does this PR close?
Closes #6370.

# Rationale for this change
1. Natively operate on StringViewArray without having to convert first to StringArray
2. (Potentially) take advantage of the new string view layout

# What changes are included in this PR?
Introduce `regexp_is_match` and `regexp_is_match_scalar` (which can replace `regexp_is_match_utf8` and `regexp_is_match_utf8_scalar`) can perform on `StringArray` / `LargeStringArray` / `StringViewArray` arguments.

# Are there any user-facing changes?
No.

### CODE REVIEW

File: arrow-string/src/regexp.rs [hunk 1]
```diff
@@ -591,40 +604,103 @@ mod tests {
...
+        regexp_is_match_utf8::<
+            GenericStringArray<i32>,
+            GenericStringArray<i32>,
+            GenericStringArray<i32>,
+        >,
+        [true, true, true, true, false, true]
+    );
+    test_flag_utf8!(
+        test_utf8_array_regexp_is_match_insensitive_2,
+        StringViewArray::from(vec!["arrow", "arrow", "arrow", "arrow", "arrow", "arrow"]),
```
Comment (alamb) [inline arrow-string/src/regexp.rs:645]: StringViewArray has special case handling for strings that are more than 12 bytes long (the string data is stored out of band in those cases)

Can you please add tests that have some strings that are longer than 12 bytes?

File: arrow-string/src/regexp.rs [hunk 2]
```diff
@@ -107,18 +116,22 @@ pub fn regexp_is_match_utf8<OffsetSize: OffsetSizeTrait>(
...
 
-/// Perform SQL `array ~ regex_array` operation on [`StringArray`] /
-/// [`LargeStringArray`] and a scalar.
+/// Perform SQL `array ~ regex_array` operation on
+/// [`StringArray`] / [`LargeStringArray`] / [`StringViewArray`] and a scalar.
 ///
 /// See the documentation on [`regexp_is_match_utf8`] for more details.
-pub fn regexp_is_match_utf8_scalar<OffsetSize: OffsetSizeTrait>(
-    array: &GenericStringArray<OffsetSize>,
+pub fn regexp_is_match_utf8_scalar<'a, S>(
```
Comment (alamb) [inline arrow-string/src/regexp.rs:127]: Unfortunately, I think this is a API change (as is the above)

**Review by alamb — COMMENTED** [review]
Thank you @tlm365  -- this is looking really nice

I wonder if you might also be willing to add StringView to the benchmarks as well, specifically

https://github.com/apache/arrow-rs/blob/704f90bbf541896387bed030e12a39be308047e8/arrow/benches/comparison_kernels.rs#L56-L62

So that if this code is changed in the future we can ensure it doesn't regress in performance

File: arrow-string/src/regexp.rs (hunk 1 above)
Comment (tlm365) [reply arrow-string/src/regexp.rs:645]: @alamb Thanks for reviewing. I'm willing to add benchmark for this one.

> Can you please add tests that have some strings that are longer than 12 bytes?

Yes, noted. I will review and update test cases for this scenario.

**Comment by tlm365** [comment]
> I wonder if you might also be willing to add StringView to the benchmarks as well, specifically
> So that if this code is changed in the future we can ensure it doesn't regress in performance

@alamb Thanks for reviewing, willing to add benchmark for this one. I will update it soon.

File: arrow-string/src/regexp.rs (hunk 2 above)
Comment (alamb) [reply arrow-string/src/regexp.rs:127]: I have an idea of how to update this PR to avoid an API change -- the reason this is important is that a breaking API change would need to wait until the next major release (Dec 2024) per the release schedule: https://github.com/apache/arrow-rs?tab=readme-ov-file#release-versioning-and-schedule

TLDR is I think if we introduced a new function like the following: 

```rust
fn regexp_is_match(
    array: &dyn Array, 
    regex_array: &dyn Array, 
    flags_array: Option<&dyn Array, >,
) -> Result<BooleanArray, ArrowError> {
..
}
``

We could then support StringView and StringArray and LargeStringArray

Comment (tlm365) [reply arrow-string/src/regexp.rs:127]: > TLDR is I think if we introduced a new function like the following:

Sounds good 👍 But why do we use `&dyn Array` for the new `regex_is_match` function instead of keeping the current implementation?

Or am I misunderstanding you? I understand that we will provide a new `regex_is_match` function, and mark the current `regex_is_match_utf8` function as:
```rust
#[deprecated(since="54.0.0", note="please use `regex_is_match` instead")]
pub fn regexp_is_match_utf8(...) { ... }
```
Is that right? 🤔

**Comment by alamb** [comment]
Here are the benchmark results (aka this PR doesn't slow down the existing implementation)

```
++ critcmp master regex-is-match-utf8
group                                                     master                                 regex-is-match-utf8
-----                                                     ------                                 -------------------
regexp_matches_utf8 scalar ends with                      1.02  1932.3±20.07µs        ? ?/sec    1.00  1898.4±17.40µs        ? ?/sec
regexp_matches_utf8 scalar starts with                    1.00  1932.1±14.07µs        ? ?/sec    1.00  1924.9±26.17µs        ? ?/sec
```

**Comment by alamb** [comment]
I am depressed about the large review backlog in this crate. We are looking for more help from the community reviewing PRs -- see https://github.com/apache/arrow-rs/issues/6418 for more

File: arrow-string/src/regexp.rs [hunk 3]
```diff
@@ -35,16 +38,64 @@ use std::sync::Arc;
...
+/// `flags_array` are optional [`StringArray`] / [`LargeStringArray`] / [`StringViewArray`] flag,
+/// which allow special search modes, such as case-insensitive and multi-line mode.
+/// See the documentation [here](https://docs.rs/regex/1.5.4/regex/#grouping-and-flags)
+/// for more information.
+///
+/// # See Also
+/// * [`regexp_is_match_scalar`] for matching a single regular expression against an array of strings
+/// * [`regexp_match`] for extracting groups from a string array based on a regular expression
+///
+/// # Example
```
Comment (alamb) [inline arrow-string/src/regexp.rs:67]: I added an example to help the migration

Screenshot 2024-09-18 at 5 40 38 PM

File: arrow-string/src/regexp.rs [hunk 4]
```diff
@@ -35,16 +38,64 @@ use std::sync::Arc;
 /// special search modes, such as case insensitive and multi-line mode.
 /// See the documentation [here](https://docs.rs/regex/1.5.4/regex/#grouping-and-flags)
 /// for more information.
+#[deprecated(since = "54.0.0", note = "please use `regex_is_match` instead")]
 pub fn regexp_is_match_utf8<OffsetSize: OffsetSizeTrait>(
     array: &GenericStringArray<OffsetSize>,
     regex_array: &GenericStringArray<OffsetSize>,
     flags_array: Option<&GenericStringArray<OffsetSize>>,
 ) -> Result<BooleanArray, ArrowError> {
+    regexp_is_match(array, regex_array, flags_array)
```
Comment (alamb) [inline arrow-string/src/regexp.rs:47]: I switched the implementation to just call the new function to avoid duplication

**Review by alamb — APPROVED** [review]
Thank you very much @tlm365  -- this looks great.

I was reviewing this PR and I had the code checked out locally, so I took the liberty of making a few changes:

1. I fixed clippy (was failing due to using deprecated functions)
1. I updated the comments / added an example to ease the transition (passing `None` as the flags argument results in type inference errors without some type help)
2. Improved the comments in general making it clearer what regexp_is_match does and how it is related to `regexp_match`

**Comment by alamb** [comment]
@tlm365  I wonder if you have a few minutes to review the changes I pushed to this PR. 

I again I am sorry about the review delays

**Comment by tlm365** [comment]
> @tlm365 I wonder if you have a few minutes to review the changes I pushed to this PR.
> 
> I again I am sorry about the review delays

@alamb Oops, thank you so much for reviewing. Sorry 🙇 I've been a little busy lately. Noted and will come back to review this weekend.

**Comment by tlm365** [comment]
> Thank you very much @tlm365 -- this looks great.
> 
> I was reviewing this PR and I had the code checked out locally, so I took the liberty of making a few changes:
> 
> 1. I fixed clippy (was failing due to using deprecated functions)
> 2. I updated the comments / added an example to ease the transition (passing `None` as the flags argument results in type inference errors without some type help)
> 3. Improved the comments in general making it clearer what regexp_is_match does and how it is related to `regexp_match`

@alamb it looks very nice 👍 thank you so much for this update! ❤️

**Review by Dandandan — APPROVED** [review]

**Comment by Dandandan** [comment]
Thanks @tlm365 and @alamb

**Comment by alamb** [comment]
We updated this PR so it was not an API change so removing the label
